internal/ethapi/transaction_args.go

```diff
@@ -139,10 +162,31 @@ func (args *TransactionArgs) ToMessage(globalGasCap uint64) types.Message {
 		log.Warn("Caller gas above allowance, capping", "requested", gas, "cap", globalGasCap)
 		gas = globalGasCap
 	}
-	gasPrice := new(big.Int)
+	var gasPrice *big.Int
 	if args.GasPrice != nil {
 		gasPrice = args.GasPrice.ToInt()
 	}
+	var feeCap *big.Int
+	if args.FeeCap != nil {
+		feeCap = args.FeeCap.ToInt()
+	}
+	var tip *big.Int
+	if args.Tip != nil {
+		tip = args.Tip.ToInt()
+	}
+	if gasPrice == nil && feeCap == nil && tip == nil {
+		log.Warn("At least one of gasPrice, feeCap, tip should be set. Defaulting to 0 gas price.")
+		gasPrice = new(big.Int)
+	}
```

Comment:
This seems wrong to me. Shouldn't it be either `gasPrice ` or both `feeCap` and `tip` ?

Reply:
I believe my reasoning for this was to allow some of the RPC methods that don't care about gas price (I think `trace_call` is an example) to not need to specify any fee mechanics. So the idea is to default `gasPrice` to `0` if no fee parameters are defined. I think this was the behaviour before 1559 txs were introduced.

Reply:
Ok, I think we need to be explicit here. These if clauses are too convoluted and hard to reason (see error below). IMHO we should handle. (legacytx, 1554tx, and if both is missing do something meaningful).

Reply:
I totally agree it's a bit convoluted now. Do you have any particular scheme in mind? My last commit just adds  more convolution into the mix